### ISSUE TITLE
einsum not fully implemented

### ISSUE THREAD

**Issue by xuancong84** [tensorflow/tensorflow#4722]
I am glad to see the newly added einsum function. The documentation claims that its usage is the same as numpy. However, it can do almost nothing as compared to numpy. For example, it only supports subscripts in the form of '_->_'. Unfortunately, even matrix transpose does not work, i.e., 'ij->ji'. 
Numpy works:

```
>>> A
array([[ 0.3828997 , -0.39114848, -0.09727838, -0.20430113],
       [ 0.48020577, -0.47122706,  0.42830791,  0.25665744],
       [-0.30885863,  0.21669025,  0.31648793,  0.22417514],
       [ 0.32505724,  0.30478035,  0.48655034,  0.20040547]])
>>> einsum('ij->ji',A)
array([[ 0.3828997 ,  0.48020577, -0.30885863,  0.32505724],
       [-0.39114848, -0.47122706,  0.21669025,  0.30478035],
       [-0.09727838,  0.42830791,  0.31648793,  0.48655034],
       [-0.20430113,  0.25665744,  0.22417514,  0.20040547]])
```

Tensorflow does not work:

```
pseudo-code:
M_ = tf.Variable(tf.random_normal([4,4]))
N_ = tf.einsum('ij->ji',M_)              
print [M_, N_]

output:
[array([[ 0.80474716, -1.38590837, -0.3379252 , -1.24965811],
       [ 2.57852983,  0.05492432,  0.23039417, -0.74263287],
       [-2.42627382,  1.70774114,  1.19503212,  0.43006262],
       [-1.04652011, -0.32753903, -1.26430523,  0.8810069 ]], dtype=float32), 
array([[ 0.80474716, -1.38590837, -0.3379252 , -1.24965811],
       [ 2.57852983,  0.05492432,  0.23039417, -0.74263287],
       [-2.42627382,  1.70774114,  1.19503212,  0.43006262],
       [-1.04652011, -0.32753903, -1.26430523,  0.8810069 ]], dtype=float32)]
```

I want to multiply a matrix with every frame vector in every batch. Or similar operations which can be done by a simple tensor product. It seems that I still have to duplicate the matrix so many times and perform a batch_matmul, which is very inconvenient and slow and memory consuming.

I suggest tensorflow to implement either the tensordot or einsum function which can perform tensor product.

It is quite a shame that tensorflow cannot even perform basic tensor product so far :(

**Comment by tatatodd** [comment]
@xuancong84 you seem to feel strongly about the behavior of `einsum`.  We definitely welcome contributions to the codebase!

@nikhilmishra000 contributed the current implementation of `einsum` (thanks for that!) in #4378, and might have comments on the current restrictions.

**Comment by nikhilmishra000** [comment]
Hi all -- the current einsum implementation expects the output indices to be in alphabetical order (that should be fairly straightforward to fix, though). Other than that, I think it should work correctly. 

@xuancong84 Have you noticed any other bugs? Would it be easier to talk offline about any other unimplemented functionality you'd like to see?

**Comment by xuancong84** [comment]
@nikhilmishra000 Although tensor transpose does not work, one-to-many matrix multiplication is working correctly. However, I have tried and compared the following two alternatives:
```
if True:                                                                                                          
    batch_Wy1 = tf.reshape(tf.tile(weights['Wy1'],[current_batch_size,1]),[-1, hiddenSize, hiddenSize], name="Wy1")
    WyY1 = tf.batch_matmul(outputs2,batch_Wy1, name="WyY1")                                                        
else:                                                                                                              
    WyY1 = tf.einsum('ijk,kl->ijl',outputs2,weights['Wy1'])                  
```                                      
During training, the cost functions are slightly different and the results diverge gradually. Is there any justification for the numerical difference? Otherwise, there might be a potential bug. Thanks!

**Comment by nikhilmishra000** [comment]
I can verify that `tf.einsum` is working correctly as follows:
```
Av = np.random.random(a_shape)
Bv = np.random.random(b_shape)
At = tf.constant(Av)
Bt = tf.constant(Bv)

Ct = tf.einsum('ijk,kl->ijl', At, Bt)
Cv = np.einsum('ijk,kl->ijl', Av, Bv)

print np.abs(session.run(Ct) - Cv).max()
```
Which gives me an output of 0.0, and suggests that think the discrepancy is coming from elsewhere in your code. Are you perhaps initializing your variables or choosing batches randomly?

**Comment by xuancong84** [comment]
@nikhilmishra000 Yes, the forward is correct. You don't need to verify, I have verified by myself. The forward calculation of the following two gives exactly the same results up to all significant figures:
```
if True:                                                                                                          
    batch_Wy1 = tf.reshape(tf.tile(weights['Wy1'],[current_batch_size,1]),[-1, hiddenSize, hiddenSize], name="Wy1")
    WyY1 = tf.batch_matmul(outputs2,batch_Wy1, name="WyY1")                                                        
else:                                                                                                              
    WyY1 = tf.einsum('ijk,kl->ijl',outputs2,weights['Wy1'])      
```
However, if I use Adam loss, the update gives slightly different results. I didn't try other loss functions, and this is not necessarily a bug.

**Comment by jart** [comment]
Strong opinions aside, I don't think the OP is being unreasonable. If TensorFlow is offering an op that's "Like numpy.einsum" then it should ideally be roughly equivalent in terms of behavior. I just left a comment in PR #4837 suggesting that this function might be more appropriate for the contrib package.

**Comment by martinwicke** [comment]
We have a number of ops that do not fully replicate numpy behavior. That is not a problem per se, as long as the differences are clearly documented. I will encourage more complete documentation of the differences in #4837, but I don't think it needs to be in contrib -- its interface is fixed (by inheriting it from numpy). 

Of course, replicating the behavior completely is not unreasonable at all, and a good feature request. I've adjusted the tag accordingly.

**Comment by fps7806** [comment]
einsum does not support dynamic sized tensors either. If any of the matrices contains a dimension of 'None' einsum returns an error.

**Comment by dsalaj** [comment]
@fps7806 I am currently using `einsum` with dynamic batch dimension like so:
`input = tf.placeholder(dtype=tf.float32, shape=(None, n_step, n_in))`
`weights = tf.Variable(tf.truncated_normal((n_in, 1), stddev=1.0/np.sqrt(n_in)))`
`Y_predict = tf.einsum('ijk,kl->ijl', input, weights)`
It works without errors which contradicts your previous statement. Has this been patched in the meantime?

@xuancong84  I am also possibly experiencing 'bad' gradients when using `einsum`. Since the `einsum`is syntactic sugar for `matmul` I assumed there should be no problems with back propagation. But when using `einsum` even with untrainable weights (same example as above) the networks fails to learn. Is `einsum` verified to work well in backprop?

**Comment by martinwicke** [comment]
I think there's no reason einsum (being implemented as a combination of existing ops) should not support gradients. I would take a look at its implementation for clues first, but if you have a simple repro case, you should file an issue.

**Comment by bhack** [comment]
Is this still valid?

**Comment by sushreebarsa** [comment]
@xuancong84 Is this still an issue ?
Could you please have a look at this [link](https://www.tensorflow.org/api_docs/python/tf/einsum)  for tf.einsum and let us know if it helps?
Thanks!
